x/ecocredit/core/msg_bridge_receive_test.go

```diff
@@ -3,207 +3,58 @@ package core
-			getMsg: func(validMsg MsgBridgeReceive) MsgBridgeReceive {
-				validMsg.Project.Jurisdiction = ""
-				return validMsg
-			},
-			errMsg: "invalid jurisdiction",
-		},
-		{
-			name: "invalid: no start date",
-			getMsg: func(validMsg MsgBridgeReceive) MsgBridgeReceive {
-				validMsg.Batch.StartDate = nil
-				return validMsg
-			},
-			errMsg: sdkerrors.ErrInvalidRequest.Wrap("start_date is required").Error(),
-		},
-		{
-			name: "invalid: no end date",
-			getMsg: func(validMsg MsgBridgeReceive) MsgBridgeReceive {
-				validMsg.Batch.EndDate = nil
-				return validMsg
-			},
-			errMsg: sdkerrors.ErrInvalidRequest.Wrap("end_date is required").Error(),
-		},
-		{
-			name: "invalid: start date after end date",
-			getMsg: func(validMsg MsgBridgeReceive) MsgBridgeReceive {
-				before := time.Date(2020, 0, 0, 0, 0, 0, 0, time.Local)
-				after := time.Date(2021, 0, 0, 0, 0, 0, 0, time.Local)
-				validMsg.Batch.StartDate = &after
-				validMsg.Batch.EndDate = &before
-				return validMsg
-			},
-			errMsg: sdkerrors.ErrInvalidRequest.Wrap("start_date must be a time before end_date").Error(),
-		},
-		{
-			name: "invalid: project metadata length",
-			getMsg: func(validMsg MsgBridgeReceive) MsgBridgeReceive {
-				validMsg.Project.Metadata = strings.Repeat("X", MaxMetadataLength+1)
-				return validMsg
-			},
-			errMsg: sdkerrors.ErrInvalidRequest.Wrapf("project_metadata length (%d) exceeds max metadata length: %d", MaxMetadataLength+1, MaxMetadataLength).Error(),
-		},
-		{
-			name: "invalid: batch metadata length",
-			getMsg: func(validMsg MsgBridgeReceive) MsgBridgeReceive {
-				validMsg.Batch.Metadata = strings.Repeat("X", MaxMetadataLength+1)
-				return validMsg
-			},
-			errMsg: sdkerrors.ErrInvalidRequest.Wrapf("batch_metadata length (%d) exceeds max metadata length: %d", MaxMetadataLength+1, MaxMetadataLength).Error(),
-		},
-		{
-			name: "invalid: class Id",
-			getMsg: func(validMsg MsgBridgeReceive) MsgBridgeReceive {
-				validMsg.ClassId = "Foobar12345"
-				return validMsg
-			},
-			errMsg: "class ID didn't match the format",
-		},
-	}
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			t.Parallel()
-			// reset the valid message, cleaning the pointer values
-			vMsg := resetMsg()
-			msg := tt.getMsg(vMsg)
-			err := msg.ValidateBasic()
+func (s *msgBridgeReceive) TheMessage(a gocuke.DocString) {
+	s.msg = &MsgBridgeReceive{}
+	err := jsonpb.UnmarshalString(a.Content, s.msg)
+	require.NoError(s.t, err)
+}
+
+func (s *msgBridgeReceive) TheMessageIsValidated() {
+	s.checkAndSetMockValues()
 
-			if len(tt.errMsg) == 0 {
-				require.NoError(t, err)
-			} else {
-				require.ErrorContains(t, err, tt.errMsg)
-			}
-		})
+	s.err = s.msg.ValidateBasic()
+}
+
+func (s *msgBridgeReceive) ExpectTheError(a string) {
+	require.EqualError(s.t, s.err, a)
+}
+
+func (s *msgBridgeReceive) ExpectNoError() {
+	require.NoError(s.t, s.err)
+}
+
+func (s *msgBridgeReceive) checkAndSetMockValues() {
```

Comment:
im wondering if it might be better to make this type of behavior a step in the feature files? i.e.

```
GIVEN the message """some message..."
AND a project metadata string of length "X"
THEN....
```

fine either way though, just wanted to see how you feel about that approach 😄

Reply:
I like the idea. I also like having the field included in the json. I don't have strong preference. I was also thinking we might just want to include a string that exceeds the length but `foooooooooooooooooo...` looked a bit silly. I used lorum ipsum in another test but that was more fitting because it was a description for a basket. Happy to try out the suggestion here so I made the change and maybe we can reassess at another point. https://github.com/regen-network/regen-ledger/pull/1231/commits/550d76ba55597b25df8c48fe0612319231942ec0